pkg/cmd/server/origin/audit.go

```diff
@@ -0,0 +1,173 @@
+
+// WithAuditTriggeredMarker is responsible for marking that the audit did actually
+// took place and fallback audit should not trigger.
+func WithAuditTriggeredMarker(handler http.Handler, contextMapper apirequest.RequestContextMapper, out io.Writer) http.Handler {
+	if out == nil {
+		return handler
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		ctx, ok := contextMapper.Get(req)
+		if !ok {
+			apiresponsewriters.InternalError(w, req, errors.New("no context found for request"))
+			return
+		}
+		contextMapper.Update(req, apirequest.WithValue(ctx, AuditTriggered, true))
+		handler.ServeHTTP(w, req)
+	})
+}
+
+// WithAuthFallbackAudit decorates a http.Handler with a fallback audit, logging
+// information only when the original one did was not triggered.
+// This needs to be used with WithAuditTriggeredMarker, which wraps the original
+// audit filter.
+func WithAuthFallbackAudit(handler http.Handler, contextMapper apirequest.RequestContextMapper, out io.Writer) http.Handler {
+	if out == nil {
+		return handler
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		respWriter := decorateResponseWriter(w, req, contextMapper, out)
+		handler.ServeHTTP(respWriter, req)
+	})
+}
+
+// decorateResponseWriter is a copy method from the upstream audit, adapted
+// to work with the fallback audit mechanism.
+func decorateResponseWriter(responseWriter http.ResponseWriter, req *http.Request,
+	contextMapper apirequest.RequestContextMapper, out io.Writer) http.ResponseWriter {
+	delegate := &auditResponseWriter{
+		ResponseWriter: responseWriter,
+		req:            req,
+		contextMapper:  contextMapper,
+		out:            out,
+	}
+	// check if the ResponseWriter we're wrapping is the fancy one we need
+	// or if the basic is sufficient
+	_, cn := responseWriter.(http.CloseNotifier)
+	_, fl := responseWriter.(http.Flusher)
+	_, hj := responseWriter.(http.Hijacker)
+	if cn && fl && hj {
+		return &fancyResponseWriterDelegator{delegate}
+	}
+	return delegate
+}
+
+// getUsername returns username or information on the authn method being used.
+func getUsername(req *http.Request) string {
+	auth := strings.TrimSpace(req.Header.Get("Authorization"))
+
+	// check basic auth
+	const basicScheme string = "Basic "
+	if strings.HasPrefix(auth, basicScheme) {
+		const basic = "<basic>"
+		str, err := base64.StdEncoding.DecodeString(auth[len(basicScheme):])
+		if err != nil {
+			return basic
+		}
+		cred := strings.SplitN(string(str), ":", 2)
+		if len(cred) < 2 {
+			return basic
+		}
+		return cred[0]
+	}
+
+	// check bearer token
+	parts := strings.Split(auth, " ")
+	if len(parts) > 1 && strings.ToLower(parts[0]) == "bearer" {
+		return "<bearer>"
+	}
+
+	// other tokens
+	token := strings.TrimSpace(req.URL.Query().Get("access_token"))
+	if len(token) > 0 {
+		return "<token>"
+	}
+
+	// cert authn
+	if req.TLS != nil && len(req.TLS.PeerCertificates) > 0 {
```

Comment:
this will be confused by an auth proxy that presents a client cert and communicates user info in headers